### PR TITLE
Moved shared communication logic to CommunicationManager

### PR DESCRIPTION
**Pull request by noob247365**
- Moved shared logic into `ICommunicationManager` class
- Reorganized methods
- Other minor refactoring

### CODE REVIEW

**Comment by noob247365** [comment]
Please note I haven't tested some of these changes as I don't have a bluetooth capable device.

I will also update this thread when I have tested serial communication protocol, but feel free to test when reviewing :)

**Comment by noob247365** [comment]
Some changes made in PR #139 will need to be merged before this PR so I can fix the merge errors

**Review by noob247365 — COMMENTED** [review]
Everything looks good now, and it builds properly

**Review by danwillm — CHANGES_REQUESTED** [review]
Another good set of changes

File: src/Communication/SerialCommunicationManager.cpp [hunk 1]
```diff
@@ -50,89 +44,62 @@ bool SerialCommunicationManager::Connect() {
...
-    LogError("Error waiting for event");
-    return false;
-  }
+  DWORD dwCommEvent = 0;
+  do {
+    if (!WaitCommEvent(m_hSerial, &dwCommEvent, NULL)) {
+      LogError("Error waiting for event");
+      return false;
+    }
+  } while ((dwCommEvent & EV_RXCHAR) != EV_RXCHAR);
```
[the remark this answers is not in the document]
Comment (noob247365) [reply src/Communication/SerialCommunicationManager.cpp:80]: The code we have below (see line 92) already does that, so I don't forsee it being a problem

[the remark this answers is not in the document]
Comment (noob247365) [reply src/Communication/SerialCommunicationManager.cpp:80]: More detailed explanation for record here:

The `ReadFile` loop below will loop infinitely if there are zero bytes read. Functionally, this means it blocks until there is input. What I've done is move the blocking call to the `WaitCommEvent` method, which is designed for waiting for changes to a serial device. The code before would block infinitely until there was input, and so does the new code, just one block earlier. Should be alright to keep 👍

[the remark this answers is not in the document]
Comment (noob247365) [reply src/Communication/SerialCommunicationManager.cpp]: After further discussion, fixed this with latest commit

**Review by danwillm — APPROVED** [review]

**Review by lucas-vrtech — APPROVED** [review]

**Pull request merged**